src/joint_constraints.rs

```diff
@@ -246,69 +246,67 @@ pub fn propagate_joint_to_var_bounds(vcs: &mut VarConstraints, jcs: &JointConstr
         let mut sum_max_opt: Option<usize> = Some(0);
 
         let mut mins: Vec<(char, usize)> = Vec::with_capacity(jc.vars.len());
-        let mut maxs: Vec<(char, usize)> = Vec::with_capacity(jc.vars.len());
+        let mut maxes: Vec<(char, usize)> = Vec::with_capacity(jc.vars.len());
 
         for &v in &jc.vars {
             let (li, ui) = vcs.bounds(v);
             sum_min += li;
 
             // Track finite sum of maxes; if any is ∞, the group max is unbounded.
-            sum_max_opt = ui.and_then(|u| sum_max_opt.map(|a| a + u));
+            sum_max_opt = sum_max_opt.and_then(|a| ui.map(|u| a + u));
```

Comment:
old: only 1 `and_then`/`map` call per `ui` that's `None` (and both calls in other cases)
new: only 1 call _after_ the first `ui` that's `None`

so this should be an improvement except in cases where the list ends completely in `v` where `ui` is `None` (and even then, the new format is only one more call)